File: problems/src/java/src/ch1_arrays_and_strings/Ch1_3.java
package ch1_arrays_and_strings;

import java.util.Arrays;

public class Ch1_3 {

    public static void main(String[] args) {
        System.out.println(isPermutation("abc", "cba"));
        System.out.println(isPermutation("abc", "cbaa"));
        System.out.println(isPermutation("abc", "ccba"));
        System.out.println(isPermutation("abcc", "ccba"));
    }

    public static boolean isPermutation(String original, String permutation) {
        if (original == null || permutation == null || original.length() != permutation.length()) {
            return false;
        }
        char[] origChars = original.toCharArray();
        char[] permChars = permutation.toCharArray();
        Arrays.sort(origChars);
        Arrays.sort(permChars);
        return new String(origChars).equals(new String(permChars));
    }
}

File: problems/src/clojure/src/ch1_arrays_and_strings/ch1_3.clj
(ns ch1-arrays-and-strings.ch1-3
  (:import (java.util Arrays)))

; Finding out if the second string is a permutation of the first

; Short and sweet
(defn is-permutation? [orig perm]
  (if (or (nil? orig) (nil? perm)
          (not (= (count orig) (count perm))))
    false
    (let [origCh (sort orig)
          permCh (sort perm)]
      (= origCh permCh))))

(is-permutation? "abc" "cba")
(is-permutation? "abc" "cbaa")
(is-permutation? "abc" "ccba")
(is-permutation? "abcc" "ccba")


;;